docs/source/more_types.rst

```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+
+        def __getitem__(self, index: Union[int, slice]) -> Union[T, Sequence[T]]:
+            if isinstance(index, int):
+                # Return a T here
+            elif isinstance(index, slice):
+                # Return a sequence of Ts here
+            else:
+                raise TypeError(...)
+
+There are a few additional things to note about using overloads:
+
+Runtime behavior
+----------------
+
+An overloaded function must consist of two or more overload *variants*
+followed by an *implementation*. The variants and the implementations
+must be adjacent in the code: think of them as one indivisible unit.
+
+The variant bodies must all be empty; only the implementation is allowed
+to contain code. This is because at runtime, the variants are completely
+ignored: they're overridden by the final implementation function.
+
+This means that an overloaded function is still an ordinary Python
+function! There is no automatic dispatch handling: you must manually
+handle the different types in the implementation (usually by using
+``if`` statements and ``isinstance`` checks).
+
+If you are adding an overload within a stub file, the implementation
+function should be omitted: stubs do not contain runtime logic.
+
+Type checking calls to overloads
+--------------------------------
+
+When you call an overloaded function, mypy will infer the correct
+return type using the provided variants. A call is never type checked
+against the implementation signature. This is why mypy will report calls
+like ``ip_address(4)`` as being invalid even though it matches the
+implementation signature.
+
+If multiple variants end up matching a call, mypy will, for the
+most part, select the return type corresponding to the first
+matching call. For example, consider the following program:
+
+.. code-block:: python
+
+    from typing import List, Union, overload
+
+    @overload
+    def summarize(data: List[str]) -> str: pass
+
+    @overload
+    def summarize(data: List[int]) -> int: pass
+
+    def summarize(data):
+        # ...snip...
+
+    # What is the type of 'output'? str or int?
+    output = summarize([])
+
+The ``summarize([])`` call matches both variants: an empty list could
+be either a ``List[str]`` or a ``List[int]``. In this case, mypy
+will break the tie by picking the first matching variant: ``output``
+will have an inferred type of ``str``. The implementor is responsible
+for making sure ``summarize`` breaks ties in the same way at runtime.
+
+There are a few exceptions to the "pick the first match" rule.
+For example, if multiple variants match due to an argument
+being of type ``Any``, mypy will make the inferred type also
+be ``Any``.
+
+Mypy will also prohibit you from writing overload variants that are
+inherently unsafely overlapping: for example, writing two variants
+that accept the same arguments but return different types.
+
+Type checking the implementation
+--------------------------------
```

Comment:
I would rename this to "Type checking the definition" and include more info here. In particular I would show two examples: one showing the "unsafely overlaps error" (e.g. with `int -> int` and `object -> str`), and one with "will be never matched" (briefly). For first example it would be great to explain the motivation with and example `x: object = 42` -> `f(x)` -> `BOOM!`

Reply:
Regarding this section/the surrounding text: I wonder if it might be better to just link to PEP 484, once the overload stuff is accepted there.

I'm currently trying to write the `int -> int` and `object -> str` example -- and it gets mired in the weeds pretty quickly. In particular, I'm finding it challenging to concisely explain why mypy flags this example as being unsafe yet accepts the `List[int] -> int` and `List[str] -> str` example despite the fact that we can trigger the exact same kind of error in both cases.

(More specifically, it'd be trivial for me to write an explanation that describe the design decisions behind this behavior in detail; I'm finding it non-trivial to write an explanation that's satisfying yet concise.)

Reply:
> I'm finding it non-trivial to write an explanation that's satisfying yet concise.

Yes, I know this is the hardest part.

Reply:
> I wonder if it might be better to just link to PEP 484, once the overload stuff is accepted there.

I'd rather be independent from that -- the purpose here is to be the best/clearest user documentation possible, while the goal of the PEP is to be as precise a specification as possible. Those are often conflicting goals.